File: demo/java/custom-java/src/main/java/com/baidu/bce/driver/Custom.java
package com.baidu.bce.driver;

import com.baidu.bce.models.DeviceConfig;
import com.baidu.bce.models.DriverConfig;
import com.baidu.bce.models.Job;
import com.baidu.bce.models.Property;
import com.baidu.bce.sdk.context.Context;
import com.baidu.bce.sdk.context.models.device.DeviceInfo;
import com.baidu.bce.sdk.context.models.device.Event;
import com.baidu.bce.sdk.plugin.IReport;
import com.baidu.bce.utils.L;
import lombok.Data;

import java.util.HashMap;
import java.util.Map;

@Data
public class Custom {
    private final Context ctx;
    private DriverConfig cfg;
    private Map<String, Worker> ws;
    private Map<String, Device> devs;
    private IReport report;
    private final Map<String, Thread> threads;

    public Custom(Context ctx, DriverConfig cfg, IReport report) {
        this.ctx = ctx;
        this.cfg = cfg;
        this.report = report;

        Map<String, DeviceInfo> infos = new HashMap<>();
        for (DeviceInfo info : this.ctx.getAllDevices(this.cfg.getDriverName())) {
            infos.put(info.getName(), info);
        }

        this.devs = new HashMap<>();
        for (DeviceConfig item : this.cfg.getDevices()) {
            if (infos.containsKey(item.getDevice())) {
                Device dev = new Device(item, infos.get(item.getDevice()));
                this.devs.put(item.getDevice(), dev);
            }
        }

        this.ws = new HashMap<>();
        for (Job job : this.cfg.getJobs()) {
            Device dev = this.devs.get(job.getDevice());
            if (dev != null) {
                this.ws.put(dev.getInfo().getName(), new Worker(job, dev, this.cfg.getDriverName(), this.report));
            }
        }

        this.threads = new HashMap<>();
    }

    public void start() {
        for (String key : this.ws.keySet()) {
            Thread td = new Thread(() -> {
                ws.get(key).working();
                synchronized (this.threads) {
                    this.threads.remove(key);
                }
            });
            this.threads.put(key, td);
            td.start();
        }
    }

    public void restart() {
        synchronized (this.threads) {
            for (Thread td : this.threads.values()) {
                td.interrupt();
            }
            this.threads.clear();

            for (String key : this.ws.keySet()) {
                Thread td = new Thread(() -> {
                    ws.get(key).working();
                    this.threads.remove(key);
                });
                this.threads.put(key, td);
                td.start();
            }
        }
    }

    public void stop() {
        synchronized (this.threads) {
            for (Thread td : this.threads.values()) {
                td.interrupt();
            }
            this.threads.clear();
        }
    }

    public void set(DeviceInfo info, Map<String, Object> props) {
        L.debug("Custom set (name = " + info.getName()+")");
        if (!this.ws.containsKey(info.getName())) {
            return;
        }
        Worker w = this.ws.get(info.getName());
        for (Map.Entry<String, Object> e : props.entrySet()) {
            for (Property p : w.getJob().getProperties()) {
                if (e.getKey().equals(p.getName())) {
                    Object val = e.getValue();
                    if (val instanceof Double ||
                            val instanceof Float ||
                            val instanceof Integer ||
                            val instanceof Long) {
                        w.getDevice().set(p.getIndex(), Float.parseFloat(val.toString()));
                    } else {
                        L.debug("custom set " + e.getKey() + " error");
                    }
                }
            }
        }
    }

    public void event(DeviceInfo info, Event event) {
        if (event.getType().equals(Context.TYPE_REPORT_EVENT)) {
            propertyGet(info);
        } else {
            L.debug("custom event " + info.getName());
        }
    }

    public void propertyGet(DeviceInfo info) {
        L.debug("Custom get (name = " + info.getName()+")");
        if (this.ws.containsKey(info.getName())) {
            Worker w = this.ws.get(info.getName());
            try {
                w.reportProperty();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: demo/csharp/custom-csharp/custom-csharp/Collection/Custom.cs
using DriverSDK.DmContext;
using DriverSDK.DmContext.Models;
using Buildin.Csharp.Models;
using DriverSDK.Plugin;
using DriverSDK.Utils;

namespace Buildin.Csharp.Collecton
{
    public class Custom
    {
        public Context Ctx { get; set; }
        private DriverConfig _cfg;
        private IDictionary<string, Worker> _ws;
        private IDictionary<string, Device> _devs;
        private IReport _report;
        private IDictionary<string, Thread> _threads;

        public Custom(Context ctx, DriverConfig cfg, IReport report)
        {
            Ctx = ctx;
            _cfg = cfg;
            _report = report;

            IDictionary<string, DeviceInfo> infos = new Dictionary<string, DeviceInfo>();
            foreach (DeviceInfo info in Ctx.GetAllDevices(_cfg.DriverName))
            {
                infos.Add(info.Name, info);
            }

            _devs = new Dictionary<string, Device>();
            foreach (DeviceConfig item in _cfg.Devices)
            {
                if (infos.ContainsKey(item.Device))
                {
                    Device dev = new(item, infos[item.Device]);
                    _devs.Add(item.Device, dev);
                }
            }

            _ws = new Dictionary<string, Worker>();
            foreach (Job job in _cfg.Jobs)
            {
                if (_devs.ContainsKey(job.Device))
                {
                    _ws.Add(_devs[job.Device].GetInfo().Name,
                        new Worker(job, _devs[job.Device], _cfg.DriverName, _report));
                }
            }

            _threads = new Dictionary<string, Thread>();
        }

        public void Start()
        {
            foreach (string key in _ws.Keys)
            {
                Thread td = new(() =>
                {
                    _ws[key].Working();
                    lock (_threads)
                    {
                        _threads.Remove(key);
                    }
                });
                _threads.Add(key, td);
                td.Start();
            }
        }

        public void Restart()
        {
            lock (_threads)
            {
                foreach (Thread td in _threads.Values)
                {
                    td.Interrupt();
                }
                _threads.Clear();

                foreach (string key in _ws.Keys)
                {
                    Thread td = new(() =>
                    {
                        _ws[key].Working();
                        _threads.Remove(key);
                    });
                    _threads.Add(key, td);
                    td.Start();
                }
            }
        }

        public void Stop()
        {
            lock (_threads)
            {
                foreach (Thread td in _threads.Values)
                {
                    td.Interrupt();
                }
                _threads.Clear();
            }
        }

        public void Set(DeviceInfo? info, IDictionary<string, object> props)
        {
            if (info == null || !_ws.ContainsKey(info.Name))
            {
                return;
            }
            L.Debug("Custom set (name = " + info.Name + ")");

            Worker w = _ws[info.Name];
            foreach (KeyValuePair<string, object> e in props)
            {
                foreach (Property p in w.Job.Properties)
                {
                    if (e.Key.Equals(p.Name))
                    {
                        object val = e.Value;
                        if (val is double || val is float || val is int || val is long)
                        {
                            w.Device.Set(p.Index, Convert.ToDouble(val));
                        }
                        else
                        {
                            L.Debug("custom set " + e.Key + " error");
                        }
                    }
                }
            }
        }

        public void Event(DeviceInfo info, Event? e)
        {
            if (e != null && e.Type.Equals(Context.TYPE_REPORT_EVENT))
            {
                PropertyGet(info);
            }
            else
            {
                L.Debug("custom event " + info.Name);
            }
        }

        public void PropertyGet(DeviceInfo? info)
        {
            L.Debug("Custom get");
            if (info != null && _ws.ContainsKey(info.Name))
            {
                L.Debug("Custom get (name = " + info.Name + ")");
                Worker w = _ws[info.Name];
                try
                {
                    w.ReportProperty();
                }
                catch (Exception e)
                {
                    L.Error(e.ToString());
                }
            }
        }
    }
}

